lib/analyzer/applications/java.ts

```diff
@@ -80,40 +96,90 @@ function getJarSha(jarBuffers: JarBuffer[]): JarFingerprint[] {
   });
 }
 
-function checkIfFatJarsAndUnpack(jarBuffers: JarBuffer[]): JarFingerprint[] {
-  const fingerprints: JarFingerprint[] = [];
+function unpackJarsTraverse({
+  jarBuffer,
+  jarPath,
+  desiredLevelsOfUnpacking,
+  unpackedLevels,
+  jarBuffers,
+}: {
+  jarBuffer: Buffer;
+  jarPath: string;
+  desiredLevelsOfUnpacking: number;
+  unpackedLevels: number;
+  jarBuffers: JarBuffer[];
+}): JarBuffer[] {
+  let foundJars: boolean = false;
 
-  for (const jarBuffer of jarBuffers) {
-    const nestedJars: JarBuffer[] = [];
-    const buffer = jarBuffer.digest;
-    const zip = new admzip(buffer);
+  if (unpackedLevels >= desiredLevelsOfUnpacking) {
+    // No need for further unpacking
+    jarBuffers.push({
+      location: jarPath,
+      digest: jarBuffer,
+    });
+  } else {
+    const zip = new admzip(jarBuffer);
     const zipEntries = zip.getEntries();
 
+    // technically the level should be increased only if a JAR is found, but increasing here to make
+    // sure it states the level, and not counting all the jars found, regardless of level.
+    unpackedLevels = unpackedLevels + 1;
+
     for (const zipEntry of zipEntries) {
       if (zipEntry.entryName.endsWith(".jar")) {
-        nestedJars.push({
-          location: zipEntry.entryName,
-          digest: zipEntry.getData(),
+        foundJars = true;
+        const entryData = zipEntry.getData();
+        const entryName = zipEntry.entryName;
+        jarPath = `${jarPath}/${entryName}`;
+
+        unpackJarsTraverse({
+          jarBuffer: entryData,
+          jarPath,
+          desiredLevelsOfUnpacking,
+          unpackedLevels,
+          jarBuffers,
         });
       }
     }
 
-    if (nestedJars.length > 0) {
-      fingerprints.push(
-        ...nestedJars.map((element) => {
-          return {
-            ...element,
-            digest: bufferToSha1(element.digest),
-          };
-        }),
-      );
-    } else {
-      fingerprints.push({
-        location: jarBuffer.location,
-        digest: bufferToSha1(jarBuffer.digest),
+    if (foundJars === false) {
+      jarBuffers.push({
+        location: jarPath,
+        digest: jarBuffer,
       });
+    } else {
+      // if we found jars, we unpacked them (called unpackJarsTraverse again)
```

Comment:
So parent jars (that contain other jars) are never included in results? I guess these are fat/uber jars and we will never track them as such?

If I'm getting this correctly, then perhaps we could remove dead branch and rather comment line 145, that `!foundJars` means that current jar is not fat, therefore we include it?

Reply:
> So parent jars (that contain other jars) are never included in results? I guess these are fat/uber jars and we will never track them as such?

That is correct, uber jar is essentially a wrapper so we don't care, we know that those hashes won't return anything from maven central, but we will see them in the path, later in registry 

> If I'm getting this correctly, then perhaps we could remove dead branch and rather comment line 145, that !foundJars means that current jar is not fat, therefore we include it?

That is nicer yes, I'll update!